### PR TITLE
Rciti

### PR DESCRIPTION
**Pull request by atton16**
- Fast forward rciti branch to the most recent master commit
- Migrate Enketo Survey as well as simplify and re-implement the feature to be easier to use
- Use Enketo Survey for Trip confirm
- Hide label and heatmap tabs

### CODE REVIEW

File: www/js/diary/services.js
Comment (atton16) [inline www/js/diary/services.js:419]: Need this inorder for https://github.com/e-mission/e-mission-phone/pull/798/commits/fb338f12df472f002f4b591bc315d9cd0023fef7#diff-b254f964755cdc55e579741bb46be006886e57fa24e7302e413556ab39ccb014R492 to work.

**Comment by atton16** [comment]
In this EnketoSurvey implementation, we use `json/enketoSurveyConfig.json` like so:

```json
{
  "TripConfirmSurvey": {
    "formPath": "json/trip-end-survey-multiple-select.json",
    "version": 1.2,
    "compatibleWith": 1
  }
}
```

We distinguish survey by their key name in the config file.

To launch the survey, simple call `EnketoSurveyLaunch.launch($scope, 'TripConfirmSurvey')`.

The launch method also support options (`opts`) which allows to launch the survey that is linked to `trip` or user (via `uuid`) described here: https://github.com/e-mission/e-mission-phone/commit/fb338f12df472f002f4b591bc315d9cd0023fef7#diff-cb2163d07f53ca7cd4d0b2a3940106b728d96a370f1158e50f7c06ec69b5efa6R79.

To resolve the button label for displaying, one must define the function that takes `xmlDoc` as input and return `string` in `enketo-survey-answer.js`. See example: https://github.com/e-mission/e-mission-phone/commit/fb338f12df472f002f4b591bc315d9cd0023fef7#diff-08e4bb957a2ad1e61811540af7799aca93cb431031e13cb412970f7ed3612fadR47.

Note that the label is resolved at the answer time. So it will also be saved to the database along with other data. This helps us simplify logic on `www/js/diary/list.js` file here https://github.com/e-mission/e-mission-phone/commit/fb338f12df472f002f4b591bc315d9cd0023fef7#diff-b254f964755cdc55e579741bb46be006886e57fa24e7302e413556ab39ccb014R175.

**Comment by asiripanich** [comment]
@shankari could you please review this PR? :)

**Comment by shankari** [comment]
@atton16 can you confirm that I only need to review changes after https://github.com/e-mission/e-mission-phone/pull/798/commits/610fdab4c3ffc452e988c1efc99bbfa46902ddf4?

The rest of it is just a straight pull from master, right?

**Comment by atton16** [comment]
> @atton16 can you confirm that I only need to review changes after [610fdab](https://github.com/e-mission/e-mission-phone/commit/610fdab4c3ffc452e988c1efc99bbfa46902ddf4)?
> 
> The rest of it is just a straight pull from master, right?

Yes, you only need to review after that commit.

**Comment by shankari** [comment]
@atton16 are you done with your changes? BTW, the current refactoring is at:
https://github.com/e-mission/e-mission-phone/pull/799

As part of https://github.com/e-mission/e-mission-docs/issues/674 and long-term maintainability, I plan to think through how to create a similar directive for the enketo survey. So it would be good if this is as close to your final version as possible.

**Comment by asiripanich** [comment]
> 
> 
> @atton16 are you done with your changes? BTW, the current refactoring is at:
> #799
> 
> As part of [e-mission/e-mission-docs#674](https://github.com/e-mission/e-mission-docs/issues/674) and long-term maintainability, I plan to think through how to create a similar directive for the enketo survey. So it would be good if this is as close to your final version as possible.

Yes, the PR is ready to be reviewed.

File: www/templates/intro/consent-text.html
Comment (shankari) [inline www/templates/intro/consent-text.html]: Do you need this given the switch to randomized token generation?

File: www/js/intro.js
Comment (shankari) [inline www/js/intro.js]: This is to auto-close the survey after completion, right?
would be good to put this into `surveylaunch.js`, since:
- others might also want to use it
- it is more likely to survive changes to the onboarding

In general, I would recommend against putting substantial functionality in `intro.js` since that tends to be heavily customized.

File: www/json/trip_confirm_options.json
Comment (shankari) [inline www/json/trip_confirm_options.json:2]: are you using the trip confirm functionality? If not, I would omit this - it just causes additional confusion and potential future merge conflicts. There is already a `trip_confirm_options.json.sample` to provide an overview.

File: www/templates/diary/list.html
Comment (shankari) [inline www/templates/diary/list.html:90]: let's avoid calling functions from the HTML. It significantly reduces performance because of the way in which angular . Create a `$scope.isAnalyzed` variable and use it instead

See https://github.com/e-mission/e-mission-phone/commit/30561b94d8691e37fc3e50b8488f5da82f6dd815 for the change which removed function based conditionals.

File: www/index.html
Comment (shankari) [inline www/index.html:49]: Note to self: we removed the ngCordova library as part of 
https://github.com/e-mission/e-mission-phone/commit/61513d92e2223aabef929c072b05110b518c0462#diff-dedb6c21fdaac9cba4da4afde881ffe5c80f80896274dbf7971d312fc6f5207d
since it is now deprecated.

File: www/js/diary/list.js
Comment (shankari) [inline www/js/diary/list.js:178]: As I said earlier, we should really create directives/modules for this functionality to make it easier to re-use without editing code.

File: www/js/diary/list.js
Comment (shankari) [inline www/js/diary/list.js:379]: Ditto. Also, we should store the result of `isAnalysed` as a variable instead of calling a function every time.

File: www/js/diary/services.js
Comment (shankari) [inline www/js/diary/services.js:1148]: thanks for fixing this annoying bug that I haven't had time to fix before.

File: www/js/diary/services.js
Comment (shankari) [reply www/js/diary/services.js:419]: I see. But why do you need to call

```
$scope.populateInputFromTimeline(tripgj, undefined, inputType, [{data: result}])
```

from `openPopover` in the first place? 

`populateInputFromTimeline`, and by extension, `getUserInputForTrip` are used for matching the input to the trip. You don't really need that because you know the match already; it is the new trip.
I don't call `populateInputFromTimeline` for the trip confirm inputs

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [inline www/js/survey/enketo-survey-service.js:137]: nit: Do you plan to store the JSON response as well? Of course, totally fine to do that as a separate enhancement.

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [inline www/js/survey/enketo-survey-service.js:141]: I see this reference to "Note that the label is resolved at the answer time."

**Review by shankari — CHANGES_REQUESTED** [review]
Overall LGTM. The Enketo code looks pretty good now! 

I have a couple of questions and nits and one real change.
Again, this really needs another round of cleanup and modularization before merging to master, but I am pretty much ready to merge to `rciti` once the comments here have been resolved so you can be unblocked for your specific project.

I haven't reviewed it `www/js/survey/*.js` in detail for correctness since I assume that it is primarily a copyover from the previous implementation.

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [reply www/js/survey/enketo-survey-service.js:137]: Having it in JSON will help with the deployer dashboard since it can be automagically displayed just like the various `confirm` fields in the trip table.

File: www/js/diary/list.js
Comment (shankari) [reply www/js/diary/list.js:379]: In fact, there is already an `isDraft` that you can use instead of creating a new function/variable.

```
tripgj.isDraft = $scope.isDraft(tripgj);
```

Comment (atton16) [reply www/js/diary/list.js:379]: Does the opposite of `isDraft` equal to **analyzed**? If so, i think it is safe to use it.

File: www/js/diary/services.js
Comment (atton16) [reply www/js/diary/services.js:419]: Inside `openPopover`, we call `EnketoSurveyLaunch.launch` to launch the survey. Then when we receive the answer, it comes in the `.then()` clause of the promise. Only that time that we can get the answer and update the label for that specific trip.

File: www/templates/intro/consent-text.html
Comment (atton16) [reply www/templates/intro/consent-text.html]: @asiripanich

File: www/json/trip_confirm_options.json
Comment (atton16) [reply www/json/trip_confirm_options.json:2]: I just put the file to avoid 404 error when debugging the phone.

File: www/js/intro.js
Comment (atton16) [reply www/js/intro.js]: I can put this bits into `surveylaunch.js`.

File: www/js/survey/enketo-survey-service.js
Comment (atton16) [reply www/js/survey/enketo-survey-service.js:120]: When the survey is launched, user can launch by adding `opts.trip`, `opts.uuid` or without any options at all.
If they use `opts.trip`, then it tries to restore answer based on the trip `start_ts` and `end_ts`.
If they use `opts.uuid`, then it tries to restore answer based on `data.uuid`.
If none `opts` is given, we fall back to line `126`.

File: www/templates/diary/list.html
Comment (atton16) [reply www/templates/diary/list.html]: Patched in https://github.com/e-mission/e-mission-phone/pull/798/commits/b57742a68746273b62a4b3bbe91b76d209d365c4.

File: www/js/survey/enketo-survey-service.js
Comment (atton16) [reply www/js/survey/enketo-survey-service.js:137]: We need to store it as  XML doc to make it compatible with Enketo library when we restore the answer. But storing JSON along with XML would be nice as you suggested. I will to this one later.

File: www/js/survey/enketo-survey-service.js
Comment (atton16) [reply www/js/survey/enketo-survey-service.js:141]: I changed the flow to make it easier to get the answer when we call `populateInputFromTimeline. Thus avoid further processing.

File: www/json/trip_confirm_options.json
Comment (shankari) [reply www/json/trip_confirm_options.json:2]: you can ignore the 404 error; there is a fallback to read the sample file in that case and everything still works

File: www/js/diary/list.js
Comment (shankari) [reply www/js/diary/list.js:382]: The implementation is not identical, but I think they should have the same effect. Feel free to look at the `isDraft` implementation!

File: www/index.html
Comment (shankari) [reply www/index.html:49]: Right, that is why it is a note **to self**

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [reply www/js/survey/enketo-survey-service.js:141]: again, this is a note to self 😄

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [reply www/js/survey/enketo-survey-service.js:120]: I understand that.

However, the code here is

```
        if (answer) {
          return answer.data.xmlResponse;
        }
```

The code at line 126 is 

```
       return answer ? answer.data.xmlResponse : null;
```

They both seem to have the same functionality, so I was just wondering why you used two different code fragments. It is not a huge issue, (which is why it is labeled "nit"), but I like to follow standard patterns if possible.

File: www/js/diary/list.js
Comment (shankari) [inline www/js/diary/list.js:204]: I would prefer using `isDraft` (if possible) to avoid rciti specific code to the maximum extent

File: www/js/intro.js
Comment (shankari) [reply www/js/intro.js]: great, but please make sure it is configurable/modular. Not everybody will use Enketo (the CEO project is using Google Forms for now), and not every enketo user will want to auto-close surveys.

Comment (shankari) [reply www/js/intro.js]: LMK the ETA for this change. I'd like to figure out whether I should wait for it or not.

Comment (atton16) [reply www/js/intro.js]: Should be done in a few hours.

File: www/js/diary/services.js
Comment (shankari) [reply www/js/diary/services.js:419]: I understand how promises work. What I don't understand is why you need to call `populateInputFromTimeline`. For example, `populateInputFromTimeline` calls, in its first line:

```
        var userInput = DiaryHelper.getUserInputForTrip(tripgj, nextTripgj, inputList);
```

This iterates through all the user inputs and tries to match the trip. But in your case, you already know which trip this is the input for. So why take that additional performance hit instead of just setting the value into the trip directly?

e.g. something like

```
tripgj.userInput[inputType] = result;
```

or 

```
tripgj.userInput[inputType] = {data: result};
```

or something similar

As I said earlier, that's what I do with the current user labels (see `$scope.save` for example).

**Comment by atton16** [comment]
Ready for another round of review.

File: www/js/diary/list.js
Comment (atton16) [reply www/js/diary/list.js]: migrated to `isDraft`.

File: www/js/survey/enketo-survey-service.js
Comment (atton16) [reply www/js/survey/enketo-survey-service.js:120]: If the answer is found by checking with `trip`, I just wanted to return right away and not checking for `uuid` any further. I could not return `null` if answer is not found when checking against `trip` since we can further check on `uuid`. That is why two bits are slightly different.

File: www/templates/intro/consent-text.html
Comment (atton16) [reply www/templates/intro/consent-text.html]: Updated in this .merge commit https://github.com/e-mission/e-mission-phone/pull/798/commits/b1b8bfdca11393bd4ac91a2cdb128993a310cc41

File: www/json/trip_confirm_options.json
Comment (atton16) [reply www/json/trip_confirm_options.json:2]: Ok but how can I un-commit it?

File: www/js/intro.js
Comment (atton16) [reply www/js/intro.js]: Done

File: www/js/diary/services.js
Comment (atton16) [reply www/js/diary/services.js:412]: hmm, haven't think of that before. Thank you for pointing out. Let me try...

Comment (atton16) [reply www/js/diary/services.js:412]: I only thought that putting all ui update in the same place would be nice.

Comment (atton16) [reply www/js/diary/services.js:412]: Done https://github.com/e-mission/e-mission-phone/pull/798/commits/c5ead07d23f4e18f33bcd2bfdea7943f9ddcf534

File: www/json/trip_confirm_options.json
Comment (shankari) [reply www/json/trip_confirm_options.json:2]: `git rm --cached www/json/trip_confirm_options.json`
and then `git commit`

File: www/js/survey/enketo-survey-service.js
Comment (shankari) [reply www/js/survey/enketo-survey-service.js:120]: I think we are speaking at cross-purposes here, but this is a nit, so I am going to let it go.

**Review by shankari — CHANGES_REQUESTED** [review]
I think this is pretty much ready now.

Three issues to resolve before this merge:
- Remove the trip_confirm_options.json file
- Confirm that removal of survey functionality was intentional + add testing results on google forms
- Fix the intro after feedback from @asiripanich 

Two long-term issues that will be handled after this merge
(you may want to check out the restructuring PR first)
- store the result as JSON so it is easier to display generically in the dashboard
- create an <enketo-survey> directive (similar to the restructuring) that removes all the `if/else` survey code from `list.js`.
- Store JSON

File: www/js/survey/launch.js
Comment (atton16) [reply www/js/survey/launch.js:189]: We do not need it anymore. So I simplify the code. However we need the search param to be there always even with the blank value to make pattern matching works on our end. And the code looks nicer with out if condition.

The survey launch fine, screenshot below:
Screen Shot 2564-09-29 at 09 05 15

But you have to change the function calling to:

```javascript
        SurveyLaunch.startSurveyPrefilled(
            'https://docs.google.com/forms/d/e/1FAIpQLSeMqtl_rHJkg-lZ-9xnQQHrJyBYOQl83zpKmwXfgOUFVOOwLg/viewform', {
                uuidSearchParam: 'entry.779873496',
        });
```

File: www/json/trip_confirm_options.json
Comment (atton16) [reply www/json/trip_confirm_options.json]: I thought there is a magic command that you can scoop the file out of commit history. Anyways, done.

File: www/templates/intro/consent-text.html
Comment (atton16) [inline www/templates/intro/consent-text.html]: The consent is fixed in merge commit here.

**Review by shankari — APPROVED** [review]

**Comment by shankari** [comment]
Deployed, should be live in ~ 20 mins.

**Comment by asiripanich** [comment]
@shankari, i just emailed you a new connectionConfig file. Could you please update that before deploying it? Thanks.

**Comment by shankari** [comment]
already did update before deploying

**Pull request merged**
